### ISSUE TITLE
Add "Who build this" to `dotnet --info`

### ISSUE THREAD

**Issue by richlander** [dotnet/sdk#37725]
We're lucky to have so many organizations building .NET. However, it can be difficult to tell which build you are using. For example, it's possible on Linux to install .NET from Microsoft or from the distro and both are valid choices.

I'm thinking about something like this (see last line).

```bash
$ dotnet --info
.NET SDK:
 Version:           8.0.100
 Commit:            57efcf1350
 Workload version:  8.0.100-manifests.2d90560f

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  23.10
 OS Platform: Linux
 RID:         ubuntu.23.10-x64
 Base Path:   /usr/lib/dotnet/sdk/8.0.100/

.NET workloads installed:
 Workload version: 8.0.100-manifests.2d90560f
There are no installed workloads to display.

Host:
  Version:      8.0.0
  Architecture: x64
  Commit:       5535e31a71
  Built by: Microsoft
```

By default, that value would be blank.  The downside of this idea is that it would be easy for someone with malicious intent to say "Microsoft" for this value to trick people (and get them to run their code). I'd suggest that it is already possible to do same (in DLL resources) and by the time `dotnet --info` runs then the gig is already up.

**Comment by richlander** [comment]
@omajid

**Comment by baronfel** [comment]
One key piece of information we should add is a very clear marker of this current host/SDK is coming from an extracted tarball vs a package-based install. Over this last holiday break alone I saw a number of error reports from folks that had mixed tarball and package-based installs and so weren't having success using our tooling, and every time there was a back and forth of discovering that the user has at one point corrupted their local environment with a tarball + PATH modification.

**Comment by richlander** [comment]
How would we know that?

**Comment by baronfel** [comment]
The SDK at least has an `installtype` that can help direct the discovery, but we could also have a heuristic based on 'is the current host/SDK installed under the documented install locations or not' and then reuse the `/etc/dotnet/install_location` scheme (and registry lookup on Windows) to help answer that question. Or we could make up a new thing fit for purpose and flow that through.  We just need _something_ to make the situation user-triagable because it's a mess right now.

**Comment by richlander** [comment]
I like the `/etc/dotnet/install_location` idea. It's way too easy to unpack a tarball to the documented install location. That's what we do for containers and I often do that on my machine.

**Comment by baronfel** [comment]
@dsplaisted do we have anything like the workload installType in the SDK payload at the moment? If not, _should_ we?

**Comment by richlander** [comment]
Cool! That helps.

**Comment by dsplaisted** [comment]
> @dsplaisted do we have anything like the workload installType in the SDK payload at the moment? If not, _should_ we?

I don't think we have anything else.  It seems like maybe these should be the same thing - the install type for workloads should (I think) be the same as the install type for the SDK itself.  Right now we only differentiate between file based and MSI, but we could add more options.

**Comment by ViktorHofer** [comment]
FWIW a switch for this already exists in dotnet/sdk to enable telemetry: [`OfficialBuilder`](https://github.com/search?q=org%3Adotnet%20OfficialBuilder%20&type=code)

**Comment by richlander** [comment]
I think that does something different. That basically means "Microsoft". We need something that can result in "Red Hat", "Canonical", "Alpine Linux, ... .

**Comment by ViktorHofer** [comment]
Yes, I was just pointing out that we could pass that property in when building the VMR from source and it would get embedded into `dotnet --info` and be used for the sdk telemetry condition.

**Comment by marcpopMSFT** [comment]
OfficialBuilder was created to enable telemetry but it was done in a way that it could be reused in other places (that's why we didn't do it as a strict true/false value). So I think Viktor is right here that it could be included in the --info output (if set) and then others could choose to set that. One drawback would be if the "Builder" was not actually the value you'd want in the --info output.

**Comment by richlander** [comment]
We don't want opting into setting this value to enable telemetry.

**Comment by ViktorHofer** [comment]
Telemetry only gets enabled when the OfficialBuilder string is Microsoft.

**Comment by richlander** [comment]
Cool. So, it seems like we just need to pipe this data through to `dotnet --info` when set.

**Comment by marcpopMSFT** [comment]
The proposal is to pipe OfficialBuilder into the .version file in the installer repo build and then include that in the --info output. We should take a look at improving the format for .version and .toolsetversion.

@richlander is it ok if we only do this for SDK installs or does the ask apply to runtime-only installs when running --info.

**Comment by omajid** [comment]
Part of the reason for this proposal, I think, is to help users (or anyone helping them) identify [package installation mixups](https://learn.microsoft.com/en-us/dotnet/core/install/linux-package-mixup?pivots=os-linux-redhat). It is common in those scenarios to have the runtime from the distro and sdk from Microsoft. The host often shows only the runtime in that case. So I think there's some value in showing the runtime vendor if only the runtime is installed.

**Comment by tmds** [comment]
fyi, when I want to know if this is a Microsoft/distro build and installed from packages/out-of-band, I look at the `RID/Base Path` lines in `dotnet --info`.

```
 RID:         ubuntu.23.10-x64
 Base Path:   /usr/lib/dotnet/sdk/8.0.100/
```
-> this is a distro build installed from packages

> By default, that value would be blank.

I would prefer to avoid blank values by: requiring a value, having a default, or omitting the line when the value is not known.

**Comment by marcpopMSFT** [comment]
Ok, it sounds like the runtime info would be valuable as well. The output from --info for the runtime is managed in the host on the runtime side. @richlander @elinor-fung can ya'll get a bug filed on runtime to record the runtime info and include it in the --info output?

We'll use this issue to track the SDK work. For the SDK:

- Add the OfficialBuild value to the .version file
- Read that value during --info
- If present, output in the --info output

**Comment by richlander** [comment]
Runtime issue: https://github.com/dotnet/runtime/issues/98387

**Comment by agocke** [comment]
> The downside of this idea is that it would be easy for someone with malicious intent to say "Microsoft" for this value to trick people (and get them to run their code)

This has me very concerned. I don’t think this is a safe long term plan. I think we should rely on cryptographic signing for this. Cert info includes an identity. If we want to build tools to more easily see the signing info, I’m open to that.

**Comment by omajid** [comment]
> > The downside of this idea is that it would be easy for someone with malicious intent to say "Microsoft" for this value to trick people (and get them to run their code)
>
> This has me very concerned. I don’t think this is a safe long term plan. 

In this scenario, the user has already run `dotnet --info`, right? If the binary was malicious, the malicious code has already executed. How much the user trusts the output is not a really security-relevant question, is it?

**Comment by richlander** [comment]
I was also looking at this feature as informative only. It is intended to help diagnostics when the information is correct.

**Comment by agocke** [comment]
I understand the caveats, but I'm still worried that this could be misleading. Maybe we can add change the wording to something like:

```
Self-reported ID (unverified): Microsoft
```

**Comment by richlander** [comment]
I think that's much too high of a standard for us to meet. It is scary for no particular reason. Also, nothing is stopping (today or tomorrow) for the attacker to do this:

```
Official source (verified by cryptographic blockchain): Microsoft
```

An attacked could do this today to win people's trust having nothing to do with this proposal. And as Omair raises, the second you see this text, the gig is already up.

I just checked the same thing with Docker

```bash
$ docker version
Client:
 Version:           24.0.7
 API version:       1.43
 Go version:        go1.22.2
 Git commit:        24.0.7-0ubuntu4
 Built:             Wed Apr 17 20:08:25 2024
 OS/Arch:           linux/amd64
 Context:           default
```

I got my Docker from Ubuntu. I could have gotten it from Docker corp. Many distros use "commits" like that to validate that you have the correct version, particularly for CVEs. At the point information like that is wrong, you are basically in rootkit territory, need to discover that and torch your machine.

Here's similar info:

```bash
$ docker info
Client:
 Version:    24.0.7
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
```

This is similar. I guess the `buildx` component comes from Docker. It was a separate install.

**Comment by richlander** [comment]
I'm looking critically at this idea again. I think the idea is sound but it's unclear how to actually pull it off. I don't think centering this experience on the host makes sense. We don't want to host to do more and it is such a small fraction of code.

At least to start, it makes sense to center this experience on the SDK and the primary dev scenario. I also see that the `dotnet --info` output doesn't entirely make sense.

```bash
$ dotnet --info
.NET SDK 9.0.100-preview.3.24204.13 (Red Hat 9.0.100-fc39)

Environment:
 OS Name:     Mac OS X
 OS Version:  14.4
 OS Platform: Darwin
 Architecture: arm64
 RID:         osx-arm64

Components:
 Host:         9.0.0-preview.3.24172.9 (9e6ba1f68c)
 SDK:          9.0.100-preview.3.24204.13 (81f61d8290)
 MSBuild:   17.11.0-preview-24178-16+7ca3c98fa
 Workload:  9.0.100-manifests.f4f7fc79
 Base Path:   /usr/local/share/dotnet/sdk/9.0.100-preview.3.24204.13/
 
.NET workloads installed:
There are no installed workloads to display.

...
```

Microsoft builds would print (for the first line):

> .NET SDK 9.0.100-preview.3.24204.13 (Microsoft)

This approach has several advantages:

- It puts the most important information at the top 
- (finally) Gets rid of the host section.
- Puts the information in more coherent sections.
- Proposes a one line approach for version + commit.
- Aligns with the practice of other tools, per Omair's examples.

We should then add a line to docs that says that this text is provided for diagnostic purposes only and shouldn't be trusted for any purpose.

**Comment by leecow** [comment]
> Unrelated: I'm wondering if we should left justify every line. It is unclear to me if the leading spaces are buying us anything.

Seems to be following a possible convention to visually distinguish sections in lengthy output. e.g. ...

```
➜ docker info
Client:
 Version:    26.1.1
 Context:    desktop-linux
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.14.0-desktop.1
    Path:     /Users/leecow/.docker/cli-plugins/docker-buildx
```
